File: web/src/main/java/hu/bsstudio/robonaut/safetycar/SafetyCarFollowHandler.java
package hu.bsstudio.robonaut.safetycar;

import hu.bsstudio.robonaut.safetycar.model.SafetyCarFollowInformation;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class SafetyCarFollowHandler implements HandlerFunction<ServerResponse> {

    private final SafetyCarService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var safety = request.bodyToMono(SafetyCarFollowInformation.class)
            .flatMap(service::safetyCarWasFollowed);
        return ServerResponse.ok().body(safety, DetailedTeam.class);
    }
}

File: web/src/main/kotlin/hu/bsstudio/robonaut/safetycar/SafetyCarFollowHandler.kt
package hu.bsstudio.robonaut.safetycar

import hu.bsstudio.robonaut.safetycar.model.SafetyCarFollowInformation
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

class SafetyCarFollowHandler(private val service: SafetyCarService) : HandlerFunction<ServerResponse> {

    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<SafetyCarFollowInformation>()
            .flatMap(service::safetyCarWasFollowed)
            .let { detailedTeam -> ServerResponse.ok().body(detailedTeam) }
    }
}
